rasa/shared/core/domain.py

```diff
@@ -682,10 +802,26 @@ def _get_featurized_entities(self, latest_message: UserUttered) -> Set[Text]:
         )
         intent_config = self.intent_config(intent_name)
         entities = latest_message.entities
-        entity_names = set(
-            entity["entity"] for entity in entities if "entity" in entity.keys()
+
+        # If Entity Roles and Groups is used, we also need to make sure the roles and groups get featurized.
+        # We concatenate the entity label with the role/group label using a special separator to make
+        # sure that the resulting label is unique (as you can have the same role/group label for different entities).
+        entity_names = (
+            set(entity["entity"] for entity in entities if "entity" in entity.keys())
+            | set(
+                f"{entity['entity']}{ENTITY_LABEL_SEPARATOR}{entity['role']}"
+                for entity in entities
+                if "entity" in entity.keys() and "role" in entity.keys()
+            )
+            | set(
+                f"{entity['entity']}{ENTITY_LABEL_SEPARATOR}{entity['group']}"
+                for entity in entities
+                if "entity" in entity.keys() and "group" in entity.keys()
+            )
         )
 
+        # the USED_ENTITIES_KEY of an intent also contains the entity labels and the concatenated entity labels with
+        # their corresponding roles and groups labels
         wanted_entities = set(intent_config.get(USED_ENTITIES_KEY, entity_names))
 
         return entity_names.intersection(wanted_entities)
```

Comment:
Optional: consistent with using set operators above

```suggestion
        return entity_names & wanted_entities
```